mqtt/mqtt-bridge/src/pump/mod.rs

```diff
@@ -111,34 +120,106 @@ where
         let shutdown_messages = self.messages.handle();
         let messages = self.messages.run();
 
-        pin!(egress, ingress, messages);
+        pin_mut!(egress, ingress, messages);
 
-        select! {
-            _= &mut egress => {
-                error!("egress stopped unexpectedly");
-                shutdown_ingress.shutdown().await;
-                shutdown_messages.shutdown().await;
+        match future::select(messages, future::select(egress, ingress)).await {
+            Either::Left((messages, publications)) => {
+                if let Err(e) = messages {
+                    debug!(error = %e, "pump messages processor exited with error");
+                } else {
+                    debug!("pump messages processor exited");
+                }
 
-                join!(ingress, messages);
+                debug!("shutting down both ingress and egress...");
 
-            },
-            _= &mut ingress => {
-                error!("ingress stopped unexpectedly");
+                shutdown_ingress.shutdown().await;
                 shutdown_egress.shutdown().await;
-                shutdown_messages.shutdown().await;
 
-                join!(egress, messages);
-            },
-            _= &mut messages => {
-                info!("stopping pump");
+                match publications.await {
+                    Either::Left((egress, ingress)) => {
+                        if let Err(e) = egress {
+                            debug!(error = %e, "egress processing exited with error");
+                        } else {
+                            debug!("egress processing exited");
+                        }
+
+                        if let Err(e) = ingress.await {
+                            error!(error = %e, "ingress processing exited with error");
+                        } else {
+                            info!("ingress processing exited")
+                        }
+                    }
+                    Either::Right((ingress, egress)) => {
+                        if let Err(e) = ingress {
+                            debug!(error = %e, "ingress processing exited with error");
+                        } else {
+                            debug!("ingress processing exited");
+                        }
+
+                        if let Err(e) = egress.await {
+                            error!(error = %e, "egress processing exited with error");
+                        } else {
+                            info!("egress processing exited")
+                        }
```

Comment:
There might be a better way to deduplicate this logic. Helper function?

Reply:
there are types differences, I'm not sure how to dedup logic. happy to listen though :)